crucible-llvm/src/Lang/Crucible/LLVM/Functions.hs

```diff
@@ -0,0 +1,254 @@
+-- |
+-- Module           : Lang.Crucible.LLVM.Functions
+-- Description      : Register functions (CFGs and overrides)
+-- Copyright        : (c) Galois, Inc 2024
+-- License          : BSD3
+-- Maintainer       : Langston Barrett <langston@galois.com>
+-- Stability        : provisional
+--
+-- Registering functions to be used with the LLVM memory model is somewhat more
+-- complex than for other Crucible frontends, as LLVM has a notion of function
+-- pointers. Each function to be registered has to go through a few steps (the
+-- first two are common to all Crucible frontends):
+--
+-- * Create a 'FnHandle' and a 'FnState' (a translated CFG or an override)
+-- * Bind the 'FnHandle' to the 'FnState' ('OverrideSim.bindFnHandle')
+-- * Create a (global, immutable, zero-sized) allocation coressponding to the
+--   function in the 'MemImpl' ('allocFunPtr')
+-- * Register the correspondence between the function\'s name (and any aliases)
+--   and its global allocation ('registerGlobal', or via 'registerFunPtr')
+-- * Register the correspondence between the function\'s allocation and its
+--   handle ('doInstallHandle', or via 'bindLLVMHandle', 'bindLLVMCFG', or
+--   'bindLLVMFunc')
+--
+-- This module provides helpers to accomplish all of this. They\'re ordered
+-- roughly low-level/customizable to high-level/automated.
+--
+-- Perhaps surprisingly, there\'s no function that does all of the above at
+-- once. This is because there are two main places where binding functions
+-- happens:
+--
+-- * "Lang.Crucible.LLVM" registers translated CFGs, but does so lazily. In
+--   particular, this means that it initially binds the handle and allocation to
+--   a \"stub\" that, when called, will translate the actual CFG and then
+--   re-bind the handle and allocation to it.
+-- * "Lang.Crucible.LLVM.Intrinsics.Common" registers overrides, which generally
+--   apply to functions that are @declare@d but not @define@d. Thus, they
+--   already have corresponding allocations, which just need to be associated
+--   with the override.
+--
+-- Prior to these, function allocation happens in
+-- 'Lang.Crucible.LLVM.Globals.initializeMemory'.
```

Comment:
We should probably consider untangling this so that there's a single pass that (1) for each `define`, allocates and binds the initial overrides that lazily translate the CFGs and (2) for each `declare`, allocates and possibly binds any applicable overrides. I'll leave this for future work, though.